### PR TITLE
Restore 2.3.1 version check (part 2)

### PR DESCRIPTION
**Pull request by colinator27**
## Description
See https://github.com/UnderminersTeam/UndertaleModTool/pull/1696; this just moves the check to the object count unserialization stage, preventing the warning that can occur otherwise. Closes https://github.com/UnderminersTeam/UndertaleModTool/issues/1414.

## Notes
I think it's OK that this can potentially yield false positives/negatives, as the original PR says, because newer GM versions bypass this check altogether.

### CODE REVIEW

File: UndertaleModLib/UndertaleChunks.cs [hunk 1]
```diff
@@ -1522,10 +1524,10 @@ private void CheckForGMS2_3_1(UndertaleReader reader)
                 return;
             }
 
-            reader.AbsPosition = reader.ReadUInt32(); // go to the first "Point"
+            reader.AbsPosition = reader.ReadUInt32(); // Go to the first "Point"
             reader.Position += 8;
 
-            if (reader.ReadUInt32() != 0) // in 2.3 a int with the value of 0 would be set here,
+            if (reader.ReadUInt32() != 0) // In 2.3 a int with the value of 0 would be set here,
```
Comment (Miepee) [inline UndertaleModLib/UndertaleChunks.cs:1530]:
```suggestion
            if (reader.ReadUInt32() != 0) // In 2.3 an int with the value of 0 would be set here,
```

File: UndertaleModLib/UndertaleChunks.cs [hunk 2]
```diff
@@ -1522,10 +1524,10 @@ private void CheckForGMS2_3_1(UndertaleReader reader)
...
             {                             // it cannot be version 2.3 if this value isn't 0
```
Comment (Miepee) [inline UndertaleModLib/UndertaleChunks.cs:1531]:
```suggestion
            {                             // It cannot be version 2.3 if this value isn't 0
```

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 3]
```diff
@@ -141,6 +142,30 @@ public static uint UnserializeChildObjectCount(UndertaleReader reader)
         {
             reader.Position += 12;
 
+            // This check is partly duplicated from UndertaleChunks.cs, but it's necessary to handle embedded curves
+            // (For example, those in SEQN in the TS!Underswap v1.0 demo; see issue #1414)
+            if (!reader.undertaleData.IsVersionAtLeast(2, 3, 1))
+            {
+                long returnPosition = reader.AbsPosition;
+                uint numPoints = reader.ReadUInt32();
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:150]: what is a point here in this context?

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 4]
```diff
@@ -141,6 +142,30 @@ public static uint UnserializeChildObjectCount(UndertaleReader reader)
...
+                if (numPoints > 0)
+                {
+                    reader.AbsPosition += 8;
+                    if (reader.ReadUInt32() != 0) // In 2.3 a int with the value of 0 would be set here,
+                    {                             // it cannot be version 2.3 if this value isn't 0
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:155]:
```suggestion
                    {                             // It cannot be version 2.3 if this value isn't 0
```

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 5]
```diff
@@ -141,6 +142,30 @@ public static uint UnserializeChildObjectCount(UndertaleReader reader)
...
+            // This check is partly duplicated from UndertaleChunks.cs, but it's necessary to handle embedded curves
+            // (For example, those in SEQN in the TS!Underswap v1.0 demo; see issue #1414)
+            if (!reader.undertaleData.IsVersionAtLeast(2, 3, 1))
+            {
+                long returnPosition = reader.AbsPosition;
+                uint numPoints = reader.ReadUInt32();
+                if (numPoints > 0)
+                {
+                    reader.AbsPosition += 8;
+                    if (reader.ReadUInt32() != 0) // In 2.3 a int with the value of 0 would be set here,
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:154]:
```suggestion
                    if (reader.ReadUInt32() != 0) // In 2.3 an int with the value of 0 would be set here,
```

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 6]
```diff
@@ -141,6 +142,30 @@ public static uint UnserializeChildObjectCount(UndertaleReader reader)
...
+                    {                             // it cannot be version 2.3 if this value isn't 0
+                        reader.undertaleData.SetGMS2Version(2, 3, 1);
+                    }
+                    else
+                    {
+                        if (reader.ReadUInt32() == 0)                      // At all points (besides the first one)
+                            reader.undertaleData.SetGMS2Version(2, 3, 1);  // if BezierX0 equals to 0 (the above check)
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:161]:
```suggestion
                            reader.undertaleData.SetGMS2Version(2, 3, 1);  // If BezierX0 equals to 0 (the above check)
```

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs [hunk 7]
```diff
@@ -141,6 +142,30 @@ public static uint UnserializeChildObjectCount(UndertaleReader reader)
...
+                                                                           // then BezierY0 equals to 0 as well (the current check)
```
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:162]:
```suggestion
                                                                           // Then BezierY0 equals to 0 as well (the current check)
```

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs (hunk 6 above)
Comment (Miepee) [inline UndertaleModLib/Models/UndertaleAnimationCurve.cs:161]:
```suggestion
                            reader.undertaleData.SetGMS2Version(2, 3, 1);  // If BezierX0 equals to 0 (the above check)
```

**Comment by Miepee** [comment]
I thought there were more issues floating around for this one?

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs (hunk 3 above)
Comment (colinator27) [reply UndertaleModLib/Models/UndertaleAnimationCurve.cs:150]: It's the list that's called "Points", AKA points on an animation curve asset

Comment (Miepee) [reply UndertaleModLib/Models/UndertaleAnimationCurve.cs:150]: I feel like them being named something like `animPoints` would be clearer. Or `animPointCount`.

**Review by Jacky720 — APPROVED** [review]

File: UndertaleModLib/Models/UndertaleAnimationCurve.cs (hunk 3 above)
Comment (Miepee) [reply UndertaleModLib/Models/UndertaleAnimationCurve.cs:150]: IMO itd be nice if this would be addressed

Comment (colinator27) [reply UndertaleModLib/Models/UndertaleAnimationCurve.cs]: Just refactored this a little bit - should be more clear now (and also technically with one less line of duplicated code)

**Review by Miepee — APPROVED** [review]

**Pull request merged**